gc.c

```diff
@@ -5348,6 +5365,24 @@ gc_sweep_plane(rb_objspace_t *objspace, rb_heap_t *heap, uintptr_t p, bits_t bit
     } while (bitset);
 }
 
+static void
+gc_compact_finish_or_next_pool(rb_objspace_t *objspace, rb_size_pool_t *size_pool)
```

Comment:
I'm not sure why we're using a linked list for the size pools and recursively compacting size pools. This will make cross-size pool compaction difficult to implement without a significant refactor.

We should split compaction and sweeping steps so we can compact first then sweep.